src/libcollections/vec.rs

```diff
@@ -1377,9 +1377,17 @@ impl<T> ops::DerefMut for Vec<T> {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T> FromIterator<T> for Vec<T> {
     #[inline]
-    fn from_iter<I:Iterator<Item=T>>(iterator: I) -> Vec<T> {
+    fn from_iter<I:Iterator<Item=T>>(mut iterator: I) -> Vec<T> {
         let (lower, _) = iterator.size_hint();
         let mut vector = Vec::with_capacity(lower);
+        for element in iterator.by_ref().take(vector.capacity()) {
+            let len = vector.len();
+            unsafe {
+                ptr::write(vector.get_unchecked_mut(len), element);
+                vector.set_len(len + 1);
+            }
+        }
+
         for element in iterator {
```

Comment:
Another option would be a use of `fuse()`:

```diff
diff --git a/src/libcollections/vec.rs b/src/libcollections/vec.rs
index 7fcf0dd..9fd52a1 100644
--- a/src/libcollections/vec.rs
+++ b/src/libcollections/vec.rs
@@ -1380,7 +1380,9 @@ impl<T> FromIterator<T> for Vec<T> {
     fn from_iter<I:Iterator<Item=T>>(mut iterator: I) -> Vec<T> {
         let (lower, _) = iterator.size_hint();
         let mut vector = Vec::with_capacity(lower);
-        for element in iterator.by_ref().take(vector.capacity()) {
+
+        let mut i = iterator.by_ref().fuse();
+        for element in i.by_ref().take(vector.capacity()) {
             let len = vector.len();
             unsafe {
                 ptr::write(vector.get_unchecked_mut(len), element);
@@ -1388,7 +1390,7 @@ impl<T> FromIterator<T> for Vec<T> {
             }
         }
 
-        for element in iterator {
+        for element in i {
             vector.push(element)
         }
         vector
```

In this case the numbers are also not affected:

```
test heap          ... bench:        31 ns/iter (+/- 2)
test libc1         ... bench:        49 ns/iter (+/- 3)
test macro_repeat1 ... bench:        55 ns/iter (+/- 3)
test macro_repeat2 ... bench:        54 ns/iter (+/- 3)
test map_clone1    ... bench:        33 ns/iter (+/- 2)
test map_clone2    ... bench:        33 ns/iter (+/- 2)
test repeat1       ... bench:       286 ns/iter (+/- 17)
test repeat2       ... bench:       283 ns/iter (+/- 12)
```

(the repeat numbers are a bit flaky unfortunately, so that's not too surprising)